### PR TITLE
Dedup activity heartbeat timeout timer task creation

### PR DESCRIPTION
**Pull request by yycptt**

### CODE REVIEW

File: service/history/timerSequence.go [hunk 1]
```diff
@@ -163,6 +163,9 @@ func (t *timerSequenceImpl) createNextActivityTimer() (bool, error) {
 	}
 	// mark timer task mask as indication that timer task is generated
 	activityInfo.TimerTaskStatus |= timerTypeToTimerMask(firstTimerTask.timerType)
+	if firstTimerTask.timerType == timerTypeHeartbeat {
```
Comment (yux0) [inline service/history/timerSequence.go]: Add a comment about how this field use

File: service/history/timerQueueActiveProcessor.go [hunk 2]
```diff
@@ -374,8 +374,13 @@ func (t *timerQueueActiveProcessorImpl) processActivityTimeout(
...
 
 	// need to clear activity heartbeat timer task mask for new activity timer task creation
+	// NOTE: LastHeartbeatTimeoutVisibility is for deduping heartbeat timer creation as it's possible
+	// one heartbeat task was persisted multiple times with different taskIDs due to the retry logic
+	// for updating workflow execution. In that case, only one new heartbeat timeout task should be
+	// created.
 	isHeartBeatTask := task.TimeoutType == int(workflow.TimeoutTypeHeartbeat)
-	if activityInfo, ok := mutableState.GetActivityInfo(task.EventID); isHeartBeatTask && ok {
+	activityInfo, ok := mutableState.GetActivityInfo(task.EventID)
+	if isHeartBeatTask && ok && activityInfo.LastHeartbeatTimeoutVisibility <= task.VisibilityTimestamp.Unix() {
```
Comment (yux0) [inline service/history/timerQueueActiveProcessor.go]: Add some comment about why we need this check

**Comment by wxing1292** [comment]
not a blocking comment.

currently the mutable state is "versioned" using next event ID; next event ID is not a good attr protecting concurrent modification of mutable state or in this case, preventing writing the activity heartbeat > 1 times.

should have a additional attr 'dbVersion', which will +1 for each db write related to a mutable state
@yux0 @yycptt @venkat1109

File: service/history/timerQueueActiveProcessor_test.go [hunk 3]
```diff
@@ -705,6 +705,89 @@ func (s *timerQueueActiveProcessorSuite) TestProcessActivityTimeout_RetryPolicy_
...
+		int32(timerTimeout.Seconds()),
+		int32(timerTimeout.Seconds()),
+		int32(timerTimeout.Seconds()),
+		int32(heartbeatTimerTimeout.Seconds()),
+		&workflow.RetryPolicy{
+			InitialIntervalInSeconds:    common.Int32Ptr(1),
+			BackoffCoefficient:          common.Float64Ptr(1.2),
+			MaximumIntervalInSeconds:    common.Int32Ptr(5),
+			MaximumAttempts:             common.Int32Ptr(5),
+			NonRetriableErrorReasons:    []string{"（╯' - ')╯ ┻━┻ "},
```
Comment (vancexu) [inline service/history/timerQueueActiveProcessor_test.go]: ⊙.☉

**Comment by yycptt** [comment]
> not a blocking comment.
> 
> currently the mutable state is "versioned" using next event ID; next event ID is not a good attr protecting concurrent modification of mutable state or in this case, preventing writing the activity heartbeat > 1 times.
> 
> should have a additional attr 'dbVersion', which will +1 for each db write related to a mutable state
> @yux0 @yycptt @venkat1109
> 
> one more thing, plz also consider persist the heartbeat timer time along with activity (currently it is a in mem only attr)

Created a new issue: https://github.com/uber/cadence/issues/3036

**Comment by mkolodezny** [comment]
opened issue in temporal https://github.com/temporalio/temporal/issues/435 to track

**Pull request merged**